### ISSUE TITLE
Add history/audit trail to new Download functionality

### ISSUE THREAD

**Issue by jimduff-usds** [CDCgov/prime-reportstream#1308]
Bug:   The file download functionality used to track history/audit information (in the ACTION, REPORT_FILE and other tables), but that history is now missing.

1.  Talk to @mry-usds about why this is happening.  Maybe the solution to the problem is trivial?
1.  Get an Okta account from @mry-usds so you can test.  I think it should be in the `ignore` organization.
1.  Confirm the problem is occurring locally (create a file, download it, then confirm there's no history in the ACTION table)
1.   Study the actionhistory code in DownloadFunction.kt.   Ask questions of Jim as needed.  
1.  Fix the problem

(fyi @tlosbo)

**Comment by mry-usds** [comment]
DownloadFunction is OBE and needs to be removed.  Problem is that the Audit trail stuff was never copied into the HistoryFunctions code.  Move everything from DownloadFunction lines 184-195 over to HistoryFunctions::GetReportById and you'll be fine (and delete the DownloadFunction file)

**Comment by pablocaviglia** [comment]
I'm not sure if it's ok to delete the file DownloadFunction.kt given that there's an endpoint that looks like it's being used:

![image.png](https://images.zenhubusercontent.com/60bfd42b210dca0d8591aa8d/005810d3-1481-482f-995e-d24afee371ba)

And when that endpoint is called to download a file a registry gets created in the 'action' table:

![image.png](https://images.zenhubusercontent.com/60bfd42b210dca0d8591aa8d/60f3ee6e-80ce-4a98-a16e-536fc0b86c0e)

The endpoint to download is hit by the following page:

![image.png](https://images.zenhubusercontent.com/60bfd42b210dca0d8591aa8d/858a7992-b46a-4033-9233-a8594478a63e)

**Comment by jimduff-usds** [comment]
@pablocaviglia I think “OBE” is a military term - “overcome by events” - meaning DownloadFunction is no longer used, and its functionality is now in HistoryFunctions.kt.    Once you are done, you can delete DownloadFunction.kt.   Yes, its still usable and working, but I think the HistoryFunction stuff is newer/better.   @mry-usds  can you confirm?

I'd like you to do a bit more here -- I want you to understand what is going on in the code.   So please first find the code in HistoryFunction that does downloads and understand it.  Then study the way the ActionHistory related code in the old file (DownloadFunction,kt) and understand how that works.   Then create the correct ActionHistory code in the new function.

This is one of these jobs where the time spent understanding the problem will take much longer than fixing the problem!

To do proper testing, I think you’ll need @mry-usds  to create an organization in Okta called ‘ignore’ and put you in it.  Even when running locally, the code uses Okta to login.

**Comment by mry-usds** [comment]
Yes – sorry I slip into military speak every now and then…OBE = Overcome by events

DownloadFunction.kt is unused at this point and has been replaced by HistoryFunctions.kt

 

-MRY…

**Comment by pablocaviglia-baires** [comment]
@jimduff-usds I were able to make it work and register the download in the action table.

I found something else, not related to this, that we should talk about.
Right now there's a function called **checkAuthenticated** on the following location azure/HistoryFunctions.kt:477 that seems to be completely unsecured, making the system easily accessible without having credentials and allowing the user to download the reports of all the organizations.

The way used by that method to validate an authenticated user is by getting a JWT token on the http request header called 'authentication'. Then it just decodes it from base64 and starts parsing the contained data. The JWT token is very easy to create, it's full of online tools out there to create custom jwt tokens like this one:

http://jwtbuilder.jamiekurtz.com/

I created a token on that site with the organization name "la-doh":

![image](https://user-images.githubusercontent.com/71846466/121971723-d08dda80-cd4f-11eb-9098-a05e9fb4e5a8.png)

**Then I used the following static generated token and used it to download reports:**

eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE2MjM3MTIzMTcsImV4cCI6MTY1NTI0ODMxNywiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSIsIm9yZ2FuaXphdGlvbiI6WyItLWlnbm9yZSIsIlJvY2tldCJdfQ.EjhETcswT2pKhJt3bfBQLTWsbz63ZWz--m9ZSRz4zf8

**curl sample with that token to download reports:**

curl --location --request GET 'http://localhost:7071/api/history/report/' --header 'authorization: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE2MjM3MTIzMTcsImV4cCI6MTY1NTI0ODMxNywiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSIsIm9yZ2FuaXphdGlvbiI6WyItLWlnbm9yZSIsIlJvY2tldCJdfQ.EjhETcswT2pKhJt3bfBQLTWsbz63ZWz--m9ZSRz4zf8'
and I got results:

![image](https://user-images.githubusercontent.com/71846466/121972125-b7395e00-cd50-11eb-872d-438c1026db2c.png)
